src/core_plugins/kibana/public/dashboard/dashboard_state.js

```diff
@@ -10,39 +13,114 @@ import { getPersistedStateId } from 'plugins/kibana/dashboard/panel/panel_state'
 function getStateDefaults(dashboard) {
   return {
     title: dashboard.title,
+    timeRestore: dashboard.timeRestore,
     panels: dashboard.panelsJSON ? JSON.parse(dashboard.panelsJSON) : [],
     options: dashboard.optionsJSON ? JSON.parse(dashboard.optionsJSON) : {},
     uiState: dashboard.uiStateJSON ? JSON.parse(dashboard.uiStateJSON) : {},
     query: FilterUtils.getQueryFilterForDashboard(dashboard),
-    filters: FilterUtils.getFilterBarsForDashboard(dashboard)
+    filters: FilterUtils.getFilterBarsForDashboard(dashboard),
+    viewMode: dashboard.id ? DashboardViewMode.VIEW : DashboardViewMode.EDIT,
   };
 }
 
+/**
+ * Depending on how a dashboard is loaded, the filter object may contain a $$hashKey and $state that will throw
+ * off a filter comparison. This removes those variables.
+ * @param filters {Array.<Object>}
+ * @returns {Array.<Object>}
+ */
+function cleanFiltersForComparison(filters) {
+  return _.map(filters, (filter) => _.omit(filter, ['$$hashKey', '$state']));
+}
+
+/**
+ * Converts the time to a string, if it isn't already.
+ * @param time {string|Moment}
+ * @returns {string}
+ */
+function convertTimeToString(time) {
+  return typeof time === 'string' ? time : moment(time).toString();
+}
+
+/**
+ * Compares the two times, making sure they are in both compared in string format. Absolute times
+ * are sometimes stored as moment objects, but converted to strings when reloaded. Relative times are
+ * strings that are not convertible to moment objects.
+ * @param timeA {string|Moment}
+ * @param timeB {string|Moment}
+ * @returns {boolean}
+ */
+function areTimesEqual(timeA, timeB) {
+  return convertTimeToString(timeA) === convertTimeToString(timeB);
+}
+
 export class DashboardState {
   /**
+   *
    * @param dashboard {SavedDashboard}
-   * @param timefilter {Object}
-   * @param timeFilterPreviouslyStored {boolean} - I'm honestly not sure what this value
-   * means but preserving original logic after a refactor.
-   * @param quickTimeRanges {Array<Object>} An array of default time ranges that should be
-   * classified as 'quick' mode.
-   * @param AppState {Object} A class that will be used to instantiate the appState.
+   * @param AppState {AppState}
    */
-  constructor(dashboard, timefilter, timeFilterPreviouslyStored, quickTimeRanges, AppState) {
-    this.stateDefaults = getStateDefaults(dashboard);
+  constructor(dashboard, AppState) {
```

Comment:
Yea, it is ambiguous - is this referring to the object or the ui element? I really dislike our naming scheme of `saved*`.  For one, it's not even necessarily saved at this point, it might be a brand new dashboard.  I wish we had a better naming scheme for these types of objects, but I guess until we do, I'll stick with the current saved* pattern and be more specific.